### ISSUE TITLE
[Bounty: $10] enable Outrunners single screen mode

### ISSUE THREAD

**Issue by manuelmanuel** [libretro/mame2003-libretro#155]
Edit by markwkidd:
### Issue
The Outrunners dip switch to change from dual screen mode to single screen mode does not have any effect. The dip switch does work in standalone MAME 0.78.

### Resolution
This issue will be resolved and the bounty awarded when Outrunners and other titles affected by this regression can be configured for single screen mode as they can in standalone MAME 0.78.

-----

### Original post

outrunners works only in double screen, no way to switch to one screen by using the dip switch setting.
In AdvMame 0.94 the single screen mode works.

**Comment by markwkidd** [comment]
@dankcushions and @gamez-fan is there a way to change the default dip switch setting?

It seems a default of one screen would be the sensible default for any games that can run on single or multiple screens.

**Comment by gamez-fan** [comment]
PORT_DIPNAME( 0x01, 0x01, "Monitors" )

Try this change in the dip options for all the games in multi32.c it should make em default
to screen A on boot.

**Comment by markwkidd** [comment]
@gamez-fan this now looks like a regression.

**Original MAME 0.78 binary**: `outrunners.zip` boots directly into single-monitor mode as you would hope. I can access the Dip Switches via the Tab Menu, and when I change the settings, the monitor configuration changes as you would expect.

**mame2003**:
Even when I change the source to `PORT_DIPNAME( 0x01, 0x01, "Monitors" )` there is still a splitscreen display. I can enter the Dip Switches menu and I see that it now says "A only" but I also notice that I can't change the dipswitch setting to anything else from within this menu.

**Comment by Darknior** [comment]
Yes @markwkidd and i have a question about these games because i don't try, can we play two players with the two screen ?
If we can, it is better to have the choice, i many play with my children ... if the second player doesn't work, single screen is better.

**Comment by markwkidd** [comment]
@Darknoir the original MAME 0.78 gives you the option of easily switching between single player/monitor and dual monitor modes.

Currently mame2003 seems to be 'stuck' on dual monitor mode, and on top of that it displays only one player.

**Comment by gamez-fan** [comment]
Yeah PORT_DIPNAME(  0x01, 0x01, "Monitors" ) was just an idea to force the game to display
in single screen mode by removing support for the other screen options, it was worth a try
though eh..??

It seems like it's a problem with the libretro side of things that's stopping the screen display
options from working i'll check the video code for multi32 later to make sure it lines up
with MAME78

Of course we always have the option to tweak the screensize horizontally to move screen two
outta the picture hence giving us a defacto single screen view i did this for one of my own drivers
for a game called Back Fire! a while back

**Comment by Darknior** [comment]
It is a different problem, but Darius is also a double screen game, i think we can't remove one because the game is made to play with two screen. But it is really slow ... 
http://www.gametronik.com/site/rubriques/mame/Jeux/Darius%20II%20(dual%20screen)/Darius%20II%20(dual%20screen).png

**Comment by dankcushions** [comment]
Darius runs and displays fine on pi3. and pi2 according to the compatibility list. There's no problem.

**Comment by manuelmanuel** [comment]
So, no way to switch between single/dual monitor modes in outrunners... but single/dual monitor switch works in other games?

**Comment by markwkidd** [comment]
No one has had any luck finding the bug yet. @manuelmanuel you might try mame2010 for Outrunners in the meantime.

**Comment by markwkidd** [comment]
Visually, something about this bug is reminding me of the issue with vector resolution and antialiasing: https://github.com/libretro/mame2003-libretro/issues/332

It seems like, at least under certain circumstances, libretro and mame are not communicating correctly with each other about the display area/resolution. Not sure it's the same bug but just thinking out loud.

**Comment by grant2258** [comment]
Mark i done  another timing fix this might fix it check its for mame2003 though will be ported back here soon. Please compile this and see if it helps your vector problems.  https://github.com/grant2258/mame2003-libretro/commit/e3c61dac3432a88dd974ff3813ee2169d48c610f. I wont be submitting this until i fully test it @dankcushions  seems happy with it as well

**Comment by markwkidd** [comment]
@grant2258 I just tried `orunners` in mame2003-plus and the screen issue is still there.

**Comment by grant2258** [comment]
nothing has been changed as far as gfx rendering  far as i know outrunners isint a vector game is it?

**Comment by markwkidd** [comment]
I have just created a bounty and added $10 in the hopes that we might attract a developer proficient in libretro video to help with this issue which I would categorize as a regression due to some incomplete aspects of the libretro port.

@nayslayer this is the other issue I was thinking of tagging you on. Again -- no obligation. I just wanted to bring this to your attention.

**Comment by nayslayer** [comment]
Just for the record, I've figured out the problem and I'll get it implemented as part of #391. Basically, the clipping rectangles inside the output bitmap weren't accounted for, and this game (as well as `sonic`) happens to depend on those. Multi-monitor setups will be trickier, though.

**Comment by rsn8887** [comment]
I am pretty sure the issue is closely related to this line here:
https://github.com/libretro/mame2003-plus-libretro/blob/9c6d2f1ff84a2bf932a9c29da6cb9085d5b097db/src/vidhrdw/system32_vidhrdw.c#L825
```
		if (!multi32 || (multi32 && (readinputport(0xf)&(sys32sprite_monitor_select+1))>>sys32sprite_monitor_select))
			system32_draw_sprite ( bitmap, cliprect );
```
It looks to me as if this code should disable drawing all sprites on the "disabled monitors". This should speed up Outrunners if only monitor A is enabled.

Note that the input port F that is being read is this:
https://github.com/libretro/mame2003-plus-libretro/blob/2835e0369c4ada6e57f4a4064cd54e00f3f6b9df/src/drivers/multi32.c#L860
```
	PORT_START	/* port F*/
	PORT_DIPNAME( 0x03, 0x01, "Monitors" )
	PORT_DIPSETTING(    0x01, "A only" )
	PORT_DIPSETTING(    0x03, "A and B" )
	PORT_DIPSETTING(    0x02, "B only" )
```

However, once we figure out how to disable the second machine (to make the game run faster and only simulate ONE cab), then the screen size also has to be adjusted to fit the whole screen...

**Comment by nayslayer** [comment]
That's a good analysis, @rsn8887. I'll look into it. As games tend to change their clipping regions early during init, it'll probably be safe to nullify any any off-screen `drawgfx` calls too.

**Comment by rsn8887** [comment]
Since `readinputport(0xf)` is already used in the system32_vidhrdw.c code to change behavior depending on "A only" etc. settings, I also wonder if there's something in "src/inptport.c' that prevents this dipswitch to be correctly read?

https://github.com/libretro/mame2003-plus-libretro/blob/9281713b3ab705c49e296d84c79adb4cade9591e/src/inptport.c#L1

I am not setup to compile Retroarch, but it might be worth to add some kind of
`#define readinputport(0xf) 0x01` to the top of system32_vidhrdw.c to hard-code single monitor operation and see if that fixes orunners?

**Comment by rsn8887** [comment]
Update to $25 please :)

**Comment by grant2258** [comment]
Already fixed video code is in testing

**Comment by nayslayer** [comment]
This issue was resolved as a side effect of #391, independently confirmed by rsn8887 and grant2258. Thank you for your support!

**Comment by markwkidd** [comment]
Closed as resolved, thank you @nayslayer
